Recommendation Algorithm | Stage 2

elcruzo opened the pull request:
## Artist Recommendation Algorithm

### Summary

This pull request introduces a new artist recommendation algorithm to the backend of my application. The algorithm analyzes users' listening history, specifically focusing on genre, mood, and tempo tags of songs, to recommend artists. The recommendations are calculated every 3 hours using a scheduled cron job. Additionally, the pull request includes expanded tag values, a redesigned dashboard, and a new component to enhance mobile responsiveness.

### Detailed Explanation

1. **Tag Values and Weights**:
   - Defined numerical values for various genres, moods, and tempos.
   - Assigned weights to each tag type to reflect their importance in the recommendation algorithm:
     - Genre: 50%
     - Mood: 30%
     - Tempo: 20%

2. **Fetching and Processing Songs**:
   - Fetched all songs from the database, including associated user information.
   - Created a history of tags for each user, converting genres, moods, and tempos to their corresponding numerical values.

3. **Calculating User Similarities**:
   - Compared every user with every other user based on their tag histories.
   - Calculated a similarity score for each user pair by comparing their tags and adding weighted values for matches.
   - Stored user similarities if the similarity score met or exceeded a threshold of 1.0.

4. **Generating Recommendations**:
   - For each user, identified similar users and fetched their songs.
   - Counted the occurrences of each artist in the similar users' song lists.
   - Recommended the most frequently occurring artist to each user.
   - Ensured that the recommended artist's songs do not already exist in the user's tagged songs.

5. **Storing Recommendations**:
   - Saved the recommended artist for each user in the database.
   - Ensured only the most recent recommendation is displayed to the user, while previous recommendations are kept in the database for historical tracking.

6. **Scheduled Cron Job**:
   - Set up a cron job using `node-cron` to run the recommendation algorithm every 3 hours, ensuring timely and relevant recommendations.

7. **Frontend Enhancements**:
   - Expanded the tag values to include additional genres, moods, and tempos.
   - Redesigned the dashboard to improve the user interface.
   - Created a new component to enhance responsiveness on mobile devices.

### Code Changes

- **Backend Logic**:
  - Added `calculateRecommendations` function to compute and store artist recommendations.
  - Defined comprehensive tag values and weights.
  - Implemented logic to calculate user similarities and generate artist recommendations based on tag matches.

- **Cron Job Setup**:
  - Scheduled a cron job using `node-cron` to run the recommendation algorithm every 3 hours.

- **Frontend Enhancements**:
  - Updated `TaggingForm` component to include the expanded list of genres, moods, and tempos.
  - Redesigned the dashboard for a better user experience.
  - Added a new component to improve mobile responsiveness.

### Testing

- Tested the redesigned dashboard and new component for responsiveness and functionality.
- Verified the algorithm fetches and processes songs correctly (it was throwing a bug initially because in the process of redesigning the dashboard, the track artists function somehow broke), but the bug is fixed.
- Ensured user similarities are calculated accurately based on tag values and weights.
- Confirmed recommendations are generated and stored in the database as expected.
- Tested the cron job to ensure it runs the algorithm at the scheduled intervals.

### Enhancements in Recommendation Algorithm | Stage 3

- Determine the size of an artist using their follower count from the Spotify API response and their position in the top 50 global trending.
- Use this information to classify artists and enhance the recommendation algorithm to focus on niche artists.
- Continuously monitor and adjust the recommendation algorithm to improve accuracy and relevance.
- Ensured the recommended artist's songs do not already exist in the user's tagged songs.

sarahrzhang commented on Backend/routes/userRoutes.js:
```diff
@@ -441,9 +441,9 @@ router.post('/songs', authenticateJWT, async (req, res) => {
-                genre,
-                mood,
-                tempo,
+                genre: genre.toLowerCase(),
+                mood: mood.toLowerCase(),
+                tempo: mood.toLowerCase(),
```
this is the storage for the user's tag right? this might be a little nit picky but I think it's cleaner to store the raw data as is and filter it later when you are manipulating it, just a suggestion tho

elcruzo replied:
Got it. You're right, I'll fix that

sarahrzhang reviewed (changes requested):
the summary is really detailed and a nice touch! do you mind adding some screenshots or videos of the frontend changes so we can visualize things bit better?

sarahrzhang commented on Backend/utils/cronJobRecommendation.js:
```diff
@@ -24,50 +38,70 @@ async function calculateRecommendations() {
+                if (tag1.tempo === tag2.tempo) similarityScore += tagWeights.tempo;
+
+                // If similarity score is above or equal to the threshold, save the similarity
+                if (similarityScore >= threshold) {
```
isn't the max threshold 1.0 because 0.5 + 0.3 + 0.2 = 1.0? curious why you are checking >= and not just ==

elcruzo replied:
I've set the threshold at 0.75 for now to ensure I capture at least 2 similar tags effectively. However, I've been working on a more adaptable method for determining this threshold, which I plan to introduce in the next PR.

sarahrzhang replied:
gotcha! not sure if you updated it to 0.75, I still see line 56 `const threshold = 1.0;`

elcruzo replied:
Yes, I haven't pushed the changes yet

sarahrzhang commented on Backend/utils/cronJobRecommendation.js:
```diff
@@ -24,50 +38,70 @@ async function calculateRecommendations() {
-            tempo ? tagValues.tempo[tempo] : 0
-        ];
+        const tagCombination = {
+            genre: tagValues.genre[genre], // Convert genre to number
+            mood: mood ? tagValues.mood[mood] : 0, // Convert mood to number, if exists
+            tempo: tempo ? tagValues.tempo[tempo] : 0 // Convert tempo to number, if exists
```
so is genre required while mood and tempo are optional tags because they can be null? wouldn't it be easier for your algorithm to make all of them required?

elcruzo replied:
Yes, as of right now the "genre" tag is the only tag that is required but, to maintain consistency, I'm going to update the code to make all the tags required.

elcruzo replied:
Just realized this fix is going to be a big one since I have to make changes from the frontend to the database not just the backend but I'll make sure to circle back to this once I've finished updating the algorithm.

sarahrzhang commented on Backend/utils/cronJobRecommendation.js:
```diff
@@ -24,50 +38,70 @@ async function calculateRecommendations() {
+    const users = Object.keys(userTagHistory); // Get list of all users
 
+    // Compare every user with every other user
     for (let i = 0; i < users.length; i++) {
         for (let j = i + 1; j < users.length; j++) {
```
this function runtime is O(n^2) and will take a long time - not a huge deal, but if you're looking for an extra challenge, you can think about how to optimize this so maybe you don't necessarily need to compare every user with every other user to generate the best recommendations (maybe like not comparing users if they've already been compared last cron job and don't have any new tags, or skipping over users who have no tags in common, etc).

sarahrzhang reviewed (commented):
sorry my last review comments did not all save for some reason :( added a few more

elcruzo commented:
> the summary is really detailed and a nice touch! do you mind adding some screenshots or videos of the frontend changes so we can visualize things bit better?

Here is a video showing the updated frontend:
<div>
    <a href="https://www.loom.com/share/ae75739857884430be847662aa068592">
      <p>Updated Frontend Video</p>
    </a>
    <a href="https://www.loom.com/share/ae75739857884430be847662aa068592">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/ae75739857884430be847662aa068592-56dc6ea27bfdc248-full-play.gif">
    </a>
  </div>

elcruzo commented:
In this update, I added profile update functionality for users and enhanced the cron job to fetch artists' followers count, which will be instrumental in determining the popularity of an artist. Additionally, I removed a redundant HTML file front the frontend. The recommendation logic was refactored for improved efficiency by moving the actual recommendation outside the innermost loop. This recommendation is now stored in the database for every user. I also added test functions with mock data and removed redundant components in the frontend to further optimize the system.

ahdahl commented on Backend/utils/cronJobRecommendation.js:
```diff
@@ -2,114 +2,193 @@ const { PrismaClient } = require('@prisma/client');
+                    if (user1 === user2) continue; // Don't compare a user with themselves
+
+                    const user2Tags = userTagHistory[user2];
+                    let similarityScore = 0;
+
+                    // Calculate similarity score outside the innermost loop
+                    for (const tag1 of user1Tags) {
+                        for (const tag2 of user2Tags) {
+                            // Add to similarity score if tags match
+                            if (tag1.genre === tag2.genre) similarityScore += tagWeights.genre;
+                            if (tag1.mood === tag2.mood) similarityScore += tagWeights.mood;
+                            if (tag1.tempo === tag2.tempo) similarityScore += tagWeights.tempo;
+                        }
+                    }
+
+                    // If similarity score is high enough, remember it
+                    if (similarityScore >= threshold) {
+                        if (!userSimilarities[user1]) {
+                            userSimilarities[user1] = [];
+                        }
+                        if (!userSimilarities[user2]) {
+                            userSimilarities[user2] = [];
+                        }
+
+                        userSimilarities[user1].push({ userId: user2, similarityScore });
+                        userSimilarities[user2].push({ userId: user1, similarityScore });
+                    }
```
What if two users have 2 or more matching tags?
For instance:
    { id: 1, title: 'Song1', artist: 'Artist1', genre: 'pop', mood: 'happy', tempo: 'fast', userId: 1 },
    { id: 2, title: 'Song2', artist: 'Artist1', genre: 'pop', mood: 'happy', tempo: 'fast', userId: 2},
   { id: 3, title: 'Song3', artist: 'Artist2', genre: 'rock', mood: 'sad', tempo: 'medium', userId: 1 },
    { id: 2, title: 'Song4', artist: 'Artist2', genre: 'rock', mood: 'sad', tempo: 'medium', userId: 2 },

 I think the similarity score would wipe out after each loop over tagValue and tagType. userSimilarities would be {user1 => {user2, 1}, user2=>{user1, 1}}. Instead, we want to add the similarity score of the current tag to the similarity scores of tags you've already looped over. userSimilarities should be {user1 => {user2, 2}, user2=>{user1, 2}}

ahdahl replied:
Instead of 
userSimilarities[user1].push({ userId: user2, similarityScore });
I think we want     
existingSimilarityScore = userSimilarities[user1][user2]
userSimilarities[user1][user2] = existingSimilarityScore + similarityScore
You can add a test case with multiple matching tags to make sure this works

ahdahl commented on Backend/utils/cronJobRecommendation.js:
```diff
@@ -2,114 +2,191 @@ const { PrismaClient } = require('@prisma/client');
+                    let similarityScore = 0;
+
+                    // Calculate similarity score outside the innermost loop
+                    for (const tag1 of user1Tags) {
+                        for (const tag2 of user2Tags) {
+                            // Add to similarity score if tags match
+                            if (tag1.genre === tag2.genre) similarityScore += tagWeights.genre;
+                            if (tag1.mood === tag2.mood) similarityScore += tagWeights.mood;
+                            if (tag1.tempo === tag2.tempo) similarityScore += tagWeights.tempo;
+                        }
+                    }
+
+                    // If similarity score is high enough, remember it
+                    if (similarityScore >= threshold) {
+                        if (!userSimilarities[user1]) {
+                            userSimilarities[user1] = {};
+                        }
+                        if (!userSimilarities[user2]) {
+                            userSimilarities[user2] = {};
+                        }
+
+                        userSimilarities[user1][user2] = (userSimilarities[user1][user2] || 0) + similarityScore;
+                        userSimilarities[user2][user1] = (userSimilarities[user2][user1] || 0) + similarityScore;
+                    }
```
But wouldn't you still be running all this for each tag type? Meaning if you have
{ id: 1, title: 'Song1', artist: 'Artist1', genre: 'pop', mood: 'happy', tempo: 'fast', userId: 1 },
{ id: 2, title: 'Song2', artist: 'Artist1', genre: 'pop', mood: 'happy', tempo: 'fast', userId: 2},
The 2 users have 3 matching tag types, so you'd get similartiyScore = 1 but then add it three times to userSimilarities for the three tag types. So userSimilarities[user1][user2] would be 3 instead of 1

elcruzo replied:
It works now!

ahdahl reviewed (approved):
Thank you for adding some optimizations for early returns and for adding test cases to ensure the recommendation algorithm is working!

elcruzo commented:
In case anyone was wondering, I looked up the algorithm and it can be classified as User-Based Collaborative Filtering. This algorithm compares users based on their music preferences (genres, moods, tempos) and recommends artists to a user based on the preferences of similar users
